hook-janitor/src/webhooks.rs

```diff
@@ -1,55 +1,454 @@
+            r#"
+            SELECT DATE_TRUNC('hour', finished_at) AS hour,
+                (metadata->>'team_id')::bigint AS team_id,
+                (metadata->>'plugin_config_id')::bigint AS plugin_config_id,
+                count(*) as successes
+            FROM {0}
+            WHERE status = 'completed'
+              AND queue = $1
+            GROUP BY hour, team_id, plugin_config_id
+            ORDER BY hour, team_id, plugin_config_id;
+            "#,
+            self.table_name
+        );
+
+        let rows = sqlx::query_as::<_, CompletedRow>(&base_query)
+            .bind(&self.queue_name)
+            .fetch_all(&mut **tx)
+            .await
+            .map_err(|e| WebhookCleanerError::GetCompletedRowsError { error: e })?;
+
+        Ok(rows)
+    }
+
+    fn serialize_completed_rows(&self, completed_rows: Vec<CompletedRow>) -> Result<Vec<String>> {
+        let mut payloads = Vec::new();
+
+        for row in completed_rows {
+            let app_metric = AppMetric {
+                timestamp: row.hour,
+                team_id: row.team_id,
+                plugin_config_id: row.plugin_config_id,
+                job_id: None,
+                category: AppMetricCategory::Webhook,
+                successes: row.successes,
+                successes_on_retry: 0,
+                failures: 0,
+                error_uuid: None,
+                error_type: None,
+                error_details: None,
+            };
+
+            let payload = serde_json::to_string(&app_metric)
+                .map_err(|e| WebhookCleanerError::SerializeRowsError { error: e })?;
+
+            payloads.push(payload)
+        }
+
+        Ok(payloads)
+    }
+
+    async fn get_failed_rows(&self, tx: &mut Transaction<'_, Postgres>) -> Result<Vec<FailedRow>> {
+        let base_query = format!(
+            r#"
+            SELECT DATE_TRUNC('hour', finished_at) AS hour,
+                   (metadata->>'team_id')::bigint AS team_id,
+                   (metadata->>'plugin_config_id')::bigint AS plugin_config_id,
+                   CASE
+                       WHEN array_length(errors, 1) > 1
+                       THEN errors[array_length(errors, 1)]
+                       ELSE errors[1]
+                   END AS last_error,
+                   count(*) as failures
+            FROM {0}
+            WHERE status = 'failed'
+              AND queue = $1
+            GROUP BY hour, team_id, plugin_config_id, last_error
+            ORDER BY hour, team_id, plugin_config_id, last_error;
+            "#,
+            self.table_name
+        );
+
+        let rows = sqlx::query_as::<_, FailedRow>(&base_query)
+            .bind(&self.queue_name)
+            .fetch_all(&mut **tx)
+            .await
+            .map_err(|e| WebhookCleanerError::GetFailedRowsError { error: e })?;
+
+        Ok(rows)
+    }
+
+    fn serialize_failed_rows(&self, failed_rows: Vec<FailedRow>) -> Result<Vec<String>> {
+        let mut payloads = Vec::new();
```

Comment:
This method doesn't seem to access `self`, so I would personally opt to implement `Serialize` on `FailedRow` instead.

Reply:
Same thing with `serialize_completed_rows`.

Reply:
Good idea!